tests/x11/user_defined_snapshot.pm

```diff
@@ -0,0 +1,100 @@
+# SUSE's openQA tests
+#
+# Copyright © 2012-2016 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Test that zypper can refresh https repositories
+# Maintainer: Stephan Kulow <coolo@suse.com>
+
+use base "x11test";
+use strict;
+use testapi;
+use utils qw/sle_version_at_least/;
+
+sub handle_login {
+    if (get_var('DESKTOP_MINIMALX_INSTONLY')) {
+        # return at the DM and log in later into desired wm
+        return;
+    }
+    mouse_hide();
+    wait_still_screen;
+    if (get_var('DM_NEEDS_USERNAME')) {
+        type_string $username;
+    }
+    if (match_has_tag("sddm")) {
+        # make sure choose plasma5 session
+        assert_and_click "sddm-sessions-list";
+        assert_and_click "sddm-sessions-plasma5";
+        assert_and_click "sddm-password-input";
+    }
+    else {
+        send_key "ret";
+        if (!check_screen "displaymanager-password-prompt") {
+            record_soft_failure;
+            assert_screen "displaymanager-password-prompt";
+        }
+    }
+    type_string "$password";
+    send_key "ret";
+}
```

Comment:
that function is a copy-paste?

[DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself). You are calling the same code multiple times, please [extract method](http://refactoring.com/catalog/extractMethod.html).